### PR TITLE
Chapter_7

### PR DESCRIPTION
**Pull request by mirulzaei**

### CODE REVIEW

**Comment by mirulzaei** [comment]
ready

File: app/assets/stylesheets/colors.scss [hunk 1]
```diff
@@ -4,3 +4,5 @@ $lightgrey: #eaeaea;
 $darkyellow: #777;
 $darkgrey: #555;
 $blackyellow: #222;
+$white: #fff;
+$mediumlightgrey: #bbb;
```
Comment (saobv-2223) [inline app/assets/stylesheets/colors.scss]: add endline

File: app/assets/stylesheets/custom.scss [hunk 2]
```diff
@@ -97,5 +97,88 @@ footer {
...
+      margin: 0 0 30px;
+      }
+    }
+    
+    .field_with_errors {
+      @extend .has-error;
+      .form-control {
+      color: $state-danger-text;
+      }
+}
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]: add endline

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    else
+      flash[:danger] = "Please try again!"
+      render :new
+    end
+end
+
+private
+def user_params
+  params.require:user.permit :name, :email, :password, :password_confirmation
+  end
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
  end

  private
  
  def user_params
    params.require(:user).permit :name, :email, :password, :password_confirmation
  end
```

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+end
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: add endline

File: app/helpers/users_helper.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+module UsersHelper
+  # Returns the Gravatar for the given user.
+  def gravatar_for user 
+    gravatar_id = Digest::MD5::hexdigest user.email.downcase
+    gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}"
+    image_tag gravatar_url, alt: user.name, class: "gravatar"
+  end
+end
```
Comment (saobv-2223) [inline app/helpers/users_helper.rb]: add endline

File: app/views/layouts/users/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        <%= f.label :email %>
+        <%= f.email_field :email %>
+        <%= f.label :password %>
+        <%= f.password_field :password %>
+        <%= f.label :password_confirmation, t("confirmation") %>
+        <%= f.password_field :password_confirmation %>
+        <%= f.submit "Create My Account", class: "btn btn-primary" %>
+      <% end %>
+    </div>
+</div>
```
Comment (saobv-2223) [inline app/views/layouts/users/new.html.erb]: add endline

File: app/views/layouts/users/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+<div class="row">
+  <aside class="col-md-4">
+    <section class="user_info">
+      <h1>
+        <%= gravatar_for @user %>
+        <%= @user.name %>
+      </h1>
+    </section>
+  </aside>
+</div>
```
Comment (saobv-2223) [inline app/views/layouts/users/show.html.erb]: add endline

**Comment by mirulzaei** [comment]
ready

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    else
+      flash[:danger] = "Please try again!"
+      render :new
+    end
+end
+
+private
+
+def user_params
+  params.require:user.permit :name, :email, :password, :password_confirmation
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: indenting 2 spaces for those lines

File: app/helpers/users_helper.rb (hunk 5 above)
Comment (saobv-2223) [inline app/helpers/users_helper.rb]:
```suggestion
end

```

File: app/controllers/users_controller.rb (hunk 8 above)
Comment (mirulzaei) [reply app/controllers/users_controller.rb]: done

File: app/helpers/users_helper.rb (hunk 5 above)
Comment (mirulzaei) [reply app/helpers/users_helper.rb]: done

**Comment by mirulzaei** [comment]
ready

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    else
+      flash[:danger] = "Please try again!"
+      render :new
+    end
+end
+  
+private
+  
+def user_params
+  params.require:user.permit :name, :email, :password, :password_confirmation
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
  end
  
  private
  
  def user_params
    params.require:user.permit :name, :email, :password, 
  :password_confirmation
```

**Comment by mirulzaei** [comment]
ready

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+      flash[:danger] = "Please try again!"
+      render :new
+    end
+end
+  
+private
+  
+def user_params
+  params.require:user.permit :name, :email, :password, 
+  :password_confirmation
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: indenting those lines 2 spaces. I commented three times

**Comment by mirulzaei** [comment]
ready

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
+class UsersController < ApplicationController
+  def show
+    @user = User.find_by id: params[:id]
```
Comment (ngocvt-0484) [inline app/controllers/users_controller.rb:3]: please handle when @user is nil

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+      flash[:danger] = "Please try again!"
+      render :new
+    end
+  end  
+end
+
+  private
+  
+  def user_params
+    params.require:user.permit :name, :email, :password, :password_confirmation
```
Comment (ngocvt-0484) [inline app/controllers/users_controller.rb]:
```suggestion
    params.require(:user).permit :name, :email, :password, :password_confirmation
```

**Comment by mirulzaei** [comment]
ready

File: app/assets/stylesheets/sessions.scss [hunk 13]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
```
Comment (ngocvt-0484) [inline app/assets/stylesheets/sessions.scss]: remove empty file you never use

File: app/helpers/sessions_helper.rb [hunk 14]
```diff
@@ -0,0 +1,2 @@
+module SessionsHelper
```
Comment (ngocvt-0484) [inline app/helpers/sessions_helper.rb]: same
remove it

File: app/assets/stylesheets/sessions.scss (hunk 13 above)
Comment (mirulzaei) [reply app/assets/stylesheets/sessions.scss]: done

File: app/helpers/sessions_helper.rb (hunk 14 above)
Comment (mirulzaei) [reply app/helpers/sessions_helper.rb]: done

**Comment by mirulzaei** [comment]
ready

**Comment by mirulzaei** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**